Amqp/src/Akka.Streams.Amqp/Model.cs

```diff
@@ -0,0 +1,615 @@
+        {
+            declarations.ForEach(Declarations.Add);
+            return new NamedQueueSourceSettings(ConnectionSettings, Queue, Declarations, NoLocal, Exclusive, ConsumerTag,
+                Arguments);
+        }
+
+        public NamedQueueSourceSettings WithNoLocal(bool noLocal)
+        {
+            return new NamedQueueSourceSettings(ConnectionSettings, Queue, Declarations, noLocal, Exclusive, ConsumerTag,
+                Arguments);
+        }
+
+        public NamedQueueSourceSettings WithExclusive(bool exclusive)
+        {
+            return new NamedQueueSourceSettings(ConnectionSettings, Queue, Declarations, NoLocal, exclusive, ConsumerTag,
+                Arguments);
+        }
+
+        public NamedQueueSourceSettings WithConsumerTag(string consumerTag)
+        {
+            return new NamedQueueSourceSettings(ConnectionSettings, Queue, Declarations, NoLocal, Exclusive, consumerTag,
+                Arguments);
+        }
+
+        public NamedQueueSourceSettings WithArguments(params KeyValuePair<string, object>[] arguments)
+        {
+            arguments.ForEach(Arguments.Add);
+            return new NamedQueueSourceSettings(ConnectionSettings, Queue, Declarations, NoLocal, Exclusive, ConsumerTag,
+                Arguments);
+        }
+
+        public NamedQueueSourceSettings WithArguments(params (string paramName, object paramValue)[] arguments)
+        {
+            arguments.ForEach(pair=> Arguments.Add(pair.paramName,pair.paramValue));
+            return new NamedQueueSourceSettings(ConnectionSettings, Queue, Declarations, NoLocal, Exclusive, ConsumerTag,
+                Arguments);
+        }
+
+        public NamedQueueSourceSettings WithArguments(string key, object value)
+        {
+            Arguments.Add(key, value);
+            return new NamedQueueSourceSettings(ConnectionSettings, Queue, Declarations, NoLocal, Exclusive, ConsumerTag,
+                Arguments);
+        }
+
+        public override string ToString()
+        {
+            return
+                $"NamedQueueSourceSettings(ConnectionSettings={ConnectionSettings}, Queue={Queue}, Declarations={Declarations.Count}, NoLocal={NoLocal}, Exclusive={Exclusive}, Arguments={Arguments.Count})";
+        }
+    }
+
+    public sealed class TemporaryQueueSourceSettings : IAmqpSourceSettings
+    {
+        private TemporaryQueueSourceSettings(IAmqpConnectionSettings connectionSettings, string exchange,
+            IList<IDeclaration> declarations = null, string routingKey = null)
+        {
+            ConnectionSettings = connectionSettings;
+            Exchange = exchange;
+            Declarations = declarations ?? new List<IDeclaration>();
+            RoutingKey = routingKey;
+        }
+
+        public IAmqpConnectionSettings ConnectionSettings { get; }
+        public string Exchange { get; }
+        public IList<IDeclaration> Declarations { get; }
```

Comment:
Since this whole type is immutable, and `WithXXX` methods are pure, maybe it's better to use immutable data structure here? (`ImmutableList<>` or just `IReadOnlyList<>`)

Reply:
Which one do you prefer? i could replace with IReadOnlyList<>

Reply:
I doesn't really matter. `IReadOnlyList<>` it's simply more general and doesn't expose dependency on System.Collections.Immutable.